# The Task
The task is to place **products** into the **fixture**, maximizing product profit subject to the usual knapsack problem constraints:

-   **0/1 constraint:** there must be either zero or one of each product on the shelves
-   **shelf capacity constraint:** the sum of the product widths must not exceed the shelf width

`planogram.py` contains a minimal implementation.

To proceed:

1.  clone this repository
2.  **working on a new branch**, complete the implementation of `planogram.py:planogram()`.
    You are free to use your choice of additional libraries;
    make sure to update `pyproject.toml` / `requirements.txt` as necessary.
3.  get the code back to us:
    -   **private** repository on bitbucket (preferred) github, etc, or
    -   zip/tar project files and email us


In [1]:
import numpy as np
import pandas as pd
import itertools
import argparse

In [2]:
# Read in fixtures
fixture = pd.read_csv("fixture.csv")
fixture

shelf_no  shelf_width_cm
0         1            25.4
1         2            60.0
2         3            60.0
3         4            90.0
4         5            90.0

In [3]:
# Read in Products
products = pd.read_csv("products.csv")
products.head()

product_id  product_width_mm   profit
0      p6863                75   223.53
1      p5649                51  1183.93
2      p1592                65   214.08
3      p6878                46    81.36
4      p7897                40   182.46

In [4]:
# Check if datatypes are in order
print(products.info(), "\n")

print(fixture.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99 entries, 0 to 98
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   product_id        99 non-null     object 
 1   product_width_mm  99 non-null     int64  
 2   profit            99 non-null     float64
dtypes: float64(1), int64(1), object(1)
memory usage: 2.4+ KB
None 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   shelf_no        5 non-null      int64  
 1   shelf_width_cm  5 non-null      float64
dtypes: float64(1), int64(1)
memory usage: 208.0 bytes
None


# Our Solution

- We'll start shelving products according to their `profit_per_mm`, as that allows us to maximise the space on our shelves to earn the most profit.


- After that we will create a function that will permutate the order of the shelves. That function will compare the profits made by each permutation and pick the the most profitable one to generate our results. 


In [7]:
# %load planogram.py

# Function to convert shelf measurement to mm
def create_shelf_mm(fixture):
    fixture["shelf_width_mm"] = fixture["shelf_width_cm"]*10
    return fixture

# Function to arrange products according to profits per milimeter
def highest_profits_first(fixture, products):
    # Create mm shelf measurement
    create_shelf_mm(fixture) 
    
    # Sort Products by Profit_per_mm
    products["profit_per_mm"] = products["profit"]/products["product_width_mm"]
    products = products.sort_values("profit_per_mm", ascending=False)
    
    # Create new ideal_shelf column
    products["ideal_shelf"]= "None"
    
    # Set Total Profit at Zero
    total_profit = 0
    
    # Loop through all shelves
    for shelf_num in range(fixture.shape[0]):
        # Reset Master ideal shelf list
        ideal_shelf_list = []
        # Reset current shelf list
        items_on_shelf = []
        
        shelf_label = fixture["shelf_no"].iloc[shelf_num]
        space_left = fixture["shelf_width_mm"].iloc[shelf_num]
        
        # Loop through all products
        for product_num in range(products.shape[0]):
            current_shelf = products["ideal_shelf"].iloc[product_num]
            product_width = products["product_width_mm"].iloc[product_num]
            product_profit = products["profit"].iloc[product_num]
            product_id = products["product_id"].iloc[product_num]
            
            # If a product isn't shelved yet and can fit on a shelf
            if space_left >= product_width and current_shelf == "None":
                # Add product onto shelf and master list
                ideal_shelf_list.append(shelf_label)
                # Calculate amount of space left
                space_left -= product_width
                # Add product profit to total profit
                total_profit += product_profit
                # Append item onto a list
                items_on_shelf.append(product_id)
            else: 
                # If item doesn't fit, append its current status shelf onto master list
                ideal_shelf_list.append(current_shelf)
        # Update ideal_shelf column
        products["ideal_shelf"]=ideal_shelf_list
    
    return total_profit, products


# Final function to create a dataframe with most profitable placement of products
def planogram(fixture, products):
    """
    Arguments:
    - fixture :: DataFrame[["shelf_no", "shelf_width_cm"]]
    - products :: DataFrame[["product_id", "product_width_mm", "profit"]]

    Returns: DataFrame[["shelf_no", "product_id"]]
    """
    # Create a list of all possible permutations for index
    original_index = fixture.index.tolist()
    fixture_permutate = [list(x) for x in list(set(itertools.permutations(original_index)))]
    # Create empty list to nest profits and index
    profit_list = []
    #print (fixture_permutate)
    for index in fixture_permutate:
        profit = highest_profits_first(fixture.iloc[index], products)[0]
        df = highest_profits_first(fixture.iloc[index], products)[1]
        # Append a tuple of profit, index
        profit_list.append((profit, index, df))
    
    products = max(profit_list)[2]
    best_option_profit = max(profit_list)[0]
    max_possible_profit = products["profit"].sum()

    # Create a new column called "shelf_no"
    products["shelf_no"] = products["ideal_shelf"].copy()
    # Create a new df that does not half items not on shelves
    df = products[products["shelf_no"]!="None"][["shelf_no", "product_id"]]
    return df.sort_values("shelf_no")



if __name__ == "__main__":
    import numpy as np
    import pandas as pd
    import itertools

    import argparse

    parser = argparse.ArgumentParser()

    parser.add_argument(
        "--products",
        type=argparse.FileType(mode="r"),
        default="products.csv",
        help="products input file",
    )
    parser.add_argument(
        "--fixture",
        type=argparse.FileType(mode="r"),
        default="fixture.csv",
        help="fixture input file",
    )
    parser.add_argument(
        "--out", "-o", default="solution.csv", help="solution output file"
    )
    
    ### Dummy argument added here that allows function to run in Jupyter Notebooks
    ### Delete next three lines if this causes an error 
    parser.add_argument(
        '-f'
    )
    ###

    args = parser.parse_args()

    fixture = pd.read_csv(args.fixture)
    products = pd.read_csv(args.products)

    solution = planogram(fixture, products)

    solution.to_csv(args.out, index=False)

    print(
        "stats:",
        solution[["shelf_no", "product_id"]]
        .merge(fixture, on="shelf_no")
        .merge(products, on="product_id")
        .assign(n_products=1)
        .pivot_table(
            index=["shelf_no", "shelf_width_cm"],
            values=["n_products", "profit", "product_width_mm"],
            aggfunc=np.sum,
            margins=True,
        ),
        sep="\n",
    )



/Users/samuelhe/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


stats:
                         n_products  product_width_mm    profit
shelf_no shelf_width_cm                                        
1        25.4                     5               253    520.50
2        60.0                    11               600   1579.29
3        60.0                    13               598   8719.42
4        90.0                    17               897   3201.25
5        90.0                    17               890   5103.60
All                              63              3238  19124.06


# Evaluation

In [8]:
def check_final_results(fixture, products):
    """
    Arguments:
    - fixture :: DataFrame[["shelf_no", "shelf_width_cm"]]
    - products :: DataFrame[["product_id", "product_width_mm", "profit"]]

    Returns: DataFrame[["shelf_no", "product_id"]]
    """
    # Create a list of all possible permutations for index
    original_index = fixture.index.tolist()
    fixture_permutate = [list(x) for x in list(set(itertools.permutations(original_index)))]
    # Create empty list to nest profits and index
    profit_list = []
    #print (fixture_permutate)
    for index in fixture_permutate:
        profit = highest_profits_first(fixture.iloc[index], products)[0]
        df = highest_profits_first(fixture.iloc[index], products)[1]
        # Append a tuple of profit, index
        profit_list.append((profit, index, df))
    
    products = max(profit_list)[2]
    best_option_profit = max(profit_list)[0]
    max_possible_profit = products["profit"].sum()
    
    return products

In [10]:
# Create df to check results
results_check = check_final_results(fixture, products)

/Users/samuelhe/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


In [17]:
# Create
pd.set_option('display.max_rows', 99)
results_check.sort_values("profit_per_mm", ascending=False)

product_id  product_width_mm   profit  profit_per_mm ideal_shelf
90      p4833                54  1408.21      26.077963           3
8       p7122                65  1668.14      25.663692           3
1       p5649                51  1183.93      23.214314           3
53      p9129                52   862.89      16.594038           3
79      p8909                29   450.83      15.545862           3
96      p3438                56   684.51      12.223393           3
13      p6086                66   706.19      10.699848           3
26      p2722                43   421.59       9.804419           3
47      p5083                48   393.15       8.190625           3
66      p9134                41   329.42       8.034634           3
63      p7907                37   285.16       7.707027           3
22      p4825                37   272.17       7.355946           3
62      p2164                55   393.92       7.162182           5
5       p2307                51   349.70       6.856863           5
15      p1349                59   392.04       6.644746           5
57      p2039                54   354.98       6.573704           5
17      p1694                51   327.13       6.414314           5
24      p4360                57   358.70       6.292982           5
46      p4589                34   211.19       6.211471           5
23      p5072                66   357.05       5.409848           5
78      p9018                65   351.39       5.406000           5
69      p3379                55   294.84       5.360727           5
88      p4261                63   326.73       5.186190           5
18      p1419                48   243.78       5.078750           5
41      p3523                39   197.97       5.076154           5
12      p9801                54   274.04       5.074815           5
25      p3232                48   243.33       5.069375           5
65      p7196                51   244.35       4.791176           5
84      p3061                67   320.92       4.789851           4
11      p7448                52   239.36       4.603077           4
4       p7897                40   182.46       4.561500           5
31      p8120                52   227.10       4.367308           4
93      p9236                69   275.10       3.986957           4
75      p4027                46   178.52       3.880870           4
39      p2044                43   166.54       3.873023           4
91      p2233                51   179.73       3.524118           4
2       p1592                65   214.08       3.293538           4
36      p7611                35   115.24       3.292571           4
76      p5133                44   142.94       3.248636           4
72      p6114                43   138.77       3.227209           4
56      p3315                60   189.63       3.160500           4
6       p6763                60   187.41       3.123500           4
89      p8766                69   210.01       3.043623           4
59      p3866                48   143.95       2.998958           4
70      p7880                52   155.91       2.998269           4
86      p2498                54   161.76       2.995556           2
0       p6863                75   223.53       2.980400           2
35      p3474                49   141.95       2.896939           2
60      p5876                41   116.04       2.830244           4
81      p9190                19    53.23       2.801579           3
85      p1464                54   147.73       2.735741           2
21      p8346                48   130.01       2.708542           2
28      p1627                59   158.69       2.689661           2
40      p2212                35    92.87       2.653429           2
55      p4924                57   142.72       2.503860           2
87      p3405                50   116.96       2.339200           2
34      p5371                70   155.65       2.223571           2
33      p5810                49   107.42       2.192245    

In [21]:
# Maximum possible profit
results_check.profit.sum()

21732.75

### Analysis on the effectiveness of our solution. 
![](./assets/profit.png)

- **Profitability** - Our shelves contain products with a total profit of 19124.06, out of a maximum possible profit of 21732.75.


- **Free space left** - Based on the items sitting on the shelf in our solution, we have a collective 16mm of free space left. 


- **The ones that got away** -  Our function has maximised `profit_per_mm`, except for the four items shown above, which has higher `profit_per_mm` compared to the final item that made it onto a shelf, p4773. As we have only 16mm of space left, we can only reasonably consider adding two of these four items onto the shelf to increase profitability.


- **Future improvements** - Although, we'd like to believe that we've come up with the best possible scenario, the fact that we have left empty space on the shelves means that there just might be better ways of maximising profitability!
